### PR TITLE
[Merged by Bors] - feat: add Finsupp.prod/sum_eq_single

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
In several places, we unfold `Finsupp.sum` just to use `Finset.sum_eq_single`; this adds the missing lemma.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/BigOperators/Finsupp.lean [hunk 1]
```diff
@@ -212,6 +212,12 @@ theorem prod_congr {f : α →₀ M} {g1 g2 : α → M → N} (h : ∀ x ∈ f.s
 #align finsupp.prod_congr Finsupp.prod_congr
 #align finsupp.sum_congr Finsupp.sum_congr
 
+@[to_additive]
+theorem prod_eq_single {f : α →₀ M} (a : α) {g : α → M → N}
+    (h₀ : ∀ b ∈ f.support, b ≠ a → g b (f b) = 1) (h₁ : a ∉ f.support → g a (f a) = 1) :
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Finsupp.lean:217]: I think this _might_ be more useful as
```suggestion
    (h₀ : ∀ b, f b ≠ 0 → b ≠ a → g b (f b) = 1) (h₁ : f a = 0 → g a 0 = 1) :
```

File: Mathlib/Data/Finsupp/Multiset.lean [hunk 2]
```diff
@@ -110,8 +110,8 @@ theorem count_toMultiset [DecidableEq α] (f : α →₀ ℕ) (a : α) : (toMult
     _ = f.sum fun x n => n * ({x} : Multiset α).count a := by simp only [Multiset.count_nsmul]
     _ = f a * ({a} : Multiset α).count a :=
       sum_eq_single _
-        (fun a' _ H => by simp only [Multiset.count_singleton, if_false, H.symm, mul_zero]) fun H =>
-        by simp only [not_mem_support_iff.1 H, zero_mul]
+        (fun a' _ H => by simp only [Multiset.count_singleton, if_false, H.symm, mul_zero])
+        (by simp)
```
Comment (eric-wieser) [inline Mathlib/Data/Finsupp/Multiset.lean:114]: Can you leave this simp squeezed? Is it now just `zero_mul`?

Comment (Ruben-VandeVelde) [reply Mathlib/Data/Finsupp/Multiset.lean:114]: As it happens, no, even though `zero_mul` does solve it:
```lean
        (by simp only [Multiset.mem_singleton, not_true, Multiset.nodup_singleton, Multiset.count_eq_one_of_mem,
          mul_one, implies_true])
```

File: Mathlib/Data/Finsupp/Basic.lean [hunk 3]
```diff
@@ -451,11 +451,10 @@ def mapDomain (f : α → β) (v : α →₀ M) : β →₀ M :=
 
 theorem mapDomain_apply {f : α → β} (hf : Function.Injective f) (x : α →₀ M) (a : α) :
     mapDomain f x (f a) = x a := by
-  rw [mapDomain, sum_apply, sum, Finset.sum_eq_single a, single_eq_same]
+  rw [mapDomain, sum_apply, sum_eq_single a, single_eq_same]
   · intro b _ hba
     exact single_eq_of_ne (hf.ne hba)
-  · intro h
-    rw [not_mem_support_iff.1 h, single_zero, zero_apply]
+  · simp
```
Comment (eric-wieser) [inline Mathlib/Data/Finsupp/Basic.lean:457]: Can this one and the one below be kept as an `intro` / `rw`?

**Review by eric-wieser — APPROVED** [review]
bors d+

**Comment by Ruben-VandeVelde** [comment]
bors merge

**Pull request closed without merging**